core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -5682,11 +5682,21 @@ public static List mapEntries(Map<Object, Object> map) {
 
   /** Support the MAP_KEYS function. */
   public static List mapKeys(Map map) {
+    for (Object key : map.keySet()) {
```

Comment:
It would be more efficient to create the list and then check whether it contains nulls. (`ArrayList.contains` is fast.)

Reply:
On reflection, I think you should use `ImmutableList.copyOf(map.keySet())`. Immutable lists are better, and this will be a quick operation if map is an `ImmutableMap`.

Before doing this, use `map.keySet().contains(null)`. It's more efficient than your `for` loop.

Reply:
`map.keySet().contains(null)` and
```
    ImmutableList<Object> arrayList =  ImmutableList.copyOf(map.keySet());
    if(arrayList.contains(null)){
      throw new IllegalArgumentException("Cannot use null as map key");
    }
```
Is there a difference? I'm sorry I didn't understand what you meant.

Reply:
`ImmutableList.copyOf` will throw if the list contains null. You don't want that.

`ImmutableList.copyOf(map.keySet())` will be very efficient if `map` happens to be an `ImmutableMap`. You do want that.

Reply:
No, Spark will throw an exception if a null value appears

Reply:
done，thanks